compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -325,30 +333,55 @@ checkTemplate module_ template = do
-      Upgrading { past = DataRecord past, present = DataRecord present } -> checkFields origin (Upgrading {..})
-      Upgrading { past = DataVariant {}, present = DataVariant {} } -> Nothing
-      Upgrading { past = DataEnum {}, present = DataEnum {} } -> Nothing
-      Upgrading { past = DataInterface {}, present = DataInterface {} } -> Nothing
-      _ -> Just (EUpgradeError (MismatchDataConsVariety (dataTypeCon (past datatype))))
-
-checkFields :: UpgradedRecordOrigin -> Upgrading [(FieldName, Type)] -> Maybe Error
-checkFields origin fields =
-    let (deleted, existing, new) = extractDelExistNew $ HMS.fromList <$> fields
-    in
-    if not (HMS.null deleted) then
-        Just (EUpgradeError (RecordFieldsMissing origin))
+      Upgrading { _past = DataRecord _past, _present = DataRecord _present } ->
+          checkFields origin (Upgrading {..})
+      Upgrading { _past = DataVariant _past, _present = DataVariant _present } -> do
+          let upgrade = Upgrading{..}
+          (existing, new) <- checkDeleted (\_ -> EUpgradeError (VariantRemovedVariant origin)) (fmap HMS.fromList upgrade)
+          when (not (null new)) $
+              throwWithContextF present $ EUpgradeError (VariantAddedVariant origin)
+          when (not (all (foldU alphaType) existing)) $
+              throwWithContextF present $ EUpgradeError (VariantChangedVariantType origin)
+      Upgrading { _past = DataEnum _past, _present = DataEnum _present } -> do
+          let upgrade = Upgrading{..}
+          (_, new) <-
+              checkDeleted
+                (\_ -> EUpgradeError (EnumRemovedVariant origin))
+                (fmap (HMS.fromList . map (,())) upgrade)
+          when (not (null new)) $
+              throwWithContextF present $ EUpgradeError (EnumAddedVariant origin)
+      Upgrading { _past = DataInterface {}, _present = DataInterface {} } ->
+          pure ()
+      _ ->
+          throwWithContextF present (EUpgradeError (MismatchDataConsVariety (dataTypeCon (_past datatype))))
+
+checkFields :: MonadUpgrade m => UpgradedRecordOrigin -> Upgrading [(FieldName, Type)] -> m ()
+checkFields origin fields = do
+    (existing, new) <- checkDeleted (\_ -> EUpgradeError (RecordFieldsMissing origin)) (fmap HMS.fromList fields)
     -- If a field from the upgraded package has had its type changed
-    else if any matchingFieldDifferentType existing then
-        Just (EUpgradeError (RecordFieldsExistingChanged origin))
-    -- If a new field has a non-optional type
-    else if not (all newFieldOptionalType new) then
-        Just (EUpgradeError (RecordFieldsNewNonOptional origin))
-    else
-        Nothing
+    when (any matchingFieldDifferentType existing) $
+        throwWithContextF present (EUpgradeError (RecordFieldsExistingChanged origin))
+    case origin of
+      VariantConstructor{} ->
+        when (not (null new)) $
+            throwWithContextF present (EUpgradeError (VariantAddedVariantField origin))
```

Comment:
Also note that this is allowed once variant upgrading is implemented

Reply:
Yep, already tackled in the variant upgrading PR: https://github.com/digital-asset/daml/pull/18592